### ISSUE TITLE
Table feature

### ISSUE THREAD

**Issue by eribol** [fschutt/printpdf#56]
Is there any way to do table with printpdf?

**Comment by fschutt** [comment]
No. There is code for drawing rectangles, lines and text though, so you could write code to draw your table.

**Comment by dzmitry-lahoda** [comment]
seems good macro which may help https://github.com/becheran/grid

**Comment by peterpickerill** [comment]
Hi @fschutt, would you consider accepting pull requests to support this feature directly in printpdf? Do you see something like this in the roadmap going forward? I understand if you want to keep it minimal, and for my project it had everything I needed to make this work; I just had to write the code myself.

What I propose is based on a small module I've written to write table data (only text) in PDF's with printpdf. The concept is based on the grid layout in CSS; A table spans the complete width of the page, minus any margins (which you can set), and has a certain number of columns, each taking a chosen portion of the available width. Rows have a set height, are stored internally as a `Vec<String>`. The majority of the work was writing the methods and structs to calculate the coordinates in order for things to be placed correctly.

There is an `add_table()` function to loop over the rows and call `use_text()` to place it in the correct column. It can detect the end of the page automatically, and if so, create a new one, and continue until all rows are written. Unlimited tables are supported per document and borders can be added with a setting on the Table struct. This works by calculating the four points on a Cell, and using the `PdfLayerReference.add_shape()` method to add a border, making it look like a real table.

Limitations are that it doesn't have text alignment or wrapping and I haven't written any unit tests for it yet.

What do you think?

**Comment by dzmitry-lahoda** [comment]
i was able to draw table. starte with text layout left right center. than draw rectangle. than defined page limits draw into. than got table.

**Comment by peterpickerill** [comment]
I have created a [repo ](https://github.com/peterpickerill/printpdf_utils) for what I have suggested above (with the MIT license). 

It can be included in anyone's projects with the following TOML dependency;

`printpdf_utils = { git = "https://github.com/peterpickerill/printpdf_utils", branch="main", version="*"}`

Will try to get some examples for the repo soon. Any feedback is appreciated!

**Comment by fschutt** [comment]
There is currently extremely rudimentary HTML support, but it's unusable for practical purposes.

Proper support for tables / HTML / text shaping should come in 0.9.

**Comment by fschutt** [comment]
This is now implemented in master, together with regular HTML layout. Issue can be closed.
